### PR TITLE
fix #777

### LINKED ISSUE #777: Semantic DB is not portable
The semantic db currently stores absolute filenames: 
```
$ cat semanticdb
file:/Users/example-user/Projects/Meta1x/sandbox/Library.scala
[7..14): Library => _empty_.Library.
...
```
... which means that the db isn't portable across machines/directories.

Inferring or allowing for configuration of a "source root" or "sourcepath" at compile time for a particular compile unit would allow for relativizing these paths. One approach that I've seen before is to attempt to strip the package name from the source filename, and use the resulting directory as a root. But that won't be perfect, because scalac doesn't require that a source be located in a relevant package-named directory.

### PR DESCRIPTION
**Pull request by xeno-by**
/cc @stuhood

### CODE REVIEW

**Comment by xeno-by** [comment]
@olafurpg I'll finish with fat semantic databases tomorrow.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+package internal
+package semantic
+
+import java.nio.charset.Charset
+import org.scalameta.data._
+import scala.{Seq => _}
+import scala.collection.immutable.Seq
+import scala.meta.io._
+import scala.meta.internal.semantic.{schema => s}
+import scala.meta.{semantic => m}
```
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Can we try to use consistent names? I thought `m` was `meta`.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import scala.meta.{Dialect => mDialect}
```
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: What's the motivation for this rename? This gets shadowed by the `mDialect` extractor below.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        case (minput, m.Attributes(mdialect, mnames, mmessages, mdenots, msugars)) =>
+          object mDialect {
+            def unapply(mdialect: mDialect): Option[String] = {
+              val isStandard = scala.meta.Dialect.standards.exists(_._2 == mdialect)
+              if (isStandard) Some(mdialect.toString) else None
+            }
+          }
+          object mPosition {
+            def unapply(mpos: mPosition): Option[s.Range] = mpos match {
+              case scala.meta.inputs.Position.Range(`minput`, mPoint.Offset(_, sstart), mPoint.Offset(_, send)) => Some(s.Range(sstart, send))
```
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Position is fully qualified here but imported as `mPosition`. Why the difference?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import scala.meta.inputs.{Input => mInput, Position => mPosition, Point => mPoint}
+
+package object meta {
+  implicit class XtensionDatabaseSchema(mdatabase: m.Database) {
+    def toSchema(sourcepath: Sourcepath): s.Database = {
+      val sentries = mdatabase.entries.map {
+        case (minput, m.Attributes(mdialect, mnames, mmessages, mdenots, msugars)) =>
+          object mDialect {
+            def unapply(mdialect: mDialect): Option[String] = {
+              val isStandard = scala.meta.Dialect.standards.exists(_._2 == mdialect)
```
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: I'm not sure I like this `mdialect` vs. `mDialect` trick, took me a while to spot.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+              case m.Denotation(sflags, sname, sinfo) => Some(s.Denotation(sflags, sname, sinfo))
+              case _ => None
+            }
+          }
+          val spath = minput match {
+            case mInput.File(path, charset) if charset == Charset.forName("UTF-8") => sourcepath.relativize(path.toURI).getOrElse(sys.error(s"can't find $path in $sourcepath"))
+            case other => sys.error(s"bad database: unsupported input $other")
+          }
+          val sdialect = mdialect match {
+            case mDialect(sdialect) => sdialect
```
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Could we do something like `m.Dialect.standards.find(_.name == mdialect).getOrElse(sys.error)`? This whole extractor business is not helping readability IMO.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 3 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: PS, here is how this line looks in the review UI

screen shot 2017-05-11 at 10 40 45

Please try to keep lines ~100 chars

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 2 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Aah, is it to get the `object mDialect { def unapply(dialect: mDialect) }` sync between the extractor and the parameter type names? If so, then I'm not sure it's worth it.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 1 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: It is. But in this project, we don't yet have an umbrella `package object meta extends Everything`.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 2 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Fixed.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 4 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Fixed.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 3 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Fixed.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 5 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: Fixed.

**Comment by xeno-by** [comment]
@olafurpg The tests are now passing. `scalahostSbt` again ended up being super useful thanks to its integration tests of the semanticdb workflow. I suggest we postpone non-integration tests till https://github.com/scalameta/scalameta/issues/827.

File: scalahost/sbt/src/main/scala/scala/meta/scalahost/sbt/ScalahostSbtPlugin.scala [hunk 6]
```diff
@@ -60,10 +60,7 @@ object ScalahostSbtPlugin extends AutoPlugin {
     javaOptions ++= {
       if (scalametaDependencies.value.isEmpty) Nil
       else {
-        val sourcepath =
-          scalahostSourcepath.value
-            .flatMap(_.map(_.getAbsolutePath))
-            .mkString(java.io.File.pathSeparator)
+        val sourcepath = sys.props("user.dir")
```
Comment (olafurpg) [inline scalahost/sbt/src/main/scala/scala/meta/scalahost/sbt/ScalahostSbtPlugin.scala]: We may hit on performance issues in `Fragment.deep/find` with this sourcepath. For example, in the scalafmt/scalameta repos `target/` often contains (tens of) thousands of files from testkit.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 1 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: 👍

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala (hunk 2 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala]: 👍

File: scalahost/nsc/src/main/scala/scala/meta/internal/scalahost/ScalahostPipeline.scala [hunk 7]
```diff
@@ -41,9 +43,18 @@ trait ScalahostPipeline extends DatabaseOps { self: ScalahostPlugin =>
...
-        mminidb.save(outputClasspath, assumedSourcepath)
+        try {
+          if (!unit.source.file.name.endsWith(".scala")) return
+          val mminidb = m.Database(List(unit.source.toInput -> unit.toAttributes))
+          mminidb.save(outputClasspath, assumedSourcepath)
+        } catch {
+          case NonFatal(ex) =>
+            val msg = new StringWriter()
+            val path = unit.source.file.path
+            msg.write(s"failed to generate semanticdb for $path:$EOL")
```
Comment (olafurpg) [inline scalahost/nsc/src/main/scala/scala/meta/internal/scalahost/ScalahostPipeline.scala]: Nice!

File: scalahost/sbt/src/main/scala/scala/meta/scalahost/sbt/ScalahostSbtPlugin.scala (hunk 6 above)
Comment (xeno-by) [inline scalahost/sbt/src/main/scala/scala/meta/scalahost/sbt/ScalahostSbtPlugin.scala]: We can definitely improve the implementations on deep and find if the need arises. At the moment, they are intentionally very naive in order to focus the development effort.

**Pull request merged**